sdk/include/ndk/haltypes.h

```diff
@@ -39,61 +162,1375 @@ typedef enum _FIRMWARE_REENTRY
+    _Out_ PULONG OutEax,
+    _Out_ PULONG OutEbx,
+    _Out_ PULONG OutEcx,
+    _Out_ PULONG OutEdx
+);
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_LP_WRITEBACK_INVALIDATE)(
+    _In_ HAL_HV_LOGICAL_PROCESSOR_INDEX HvLpIndex
+);
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_LP_GET_MACHINE_CHECK_CONTEXT)(
+    _In_ HAL_HV_LOGICAL_PROCESSOR_INDEX HvLpIndex,
+    _Out_ UINT32 *Source,
+    _Out_ UINT64 *PartitionId,
+    _Out_ UINT32 *VpIndex
+);
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_SUSPEND_PARTITION)(
+    _In_ UINT64 PartitionId
+);
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_RESUME_PARTITION)(
+    _In_ UINT64 PartitionId
+);
+
+typedef struct _WHEA_RECOVERY_CONTEXT
+    *PWHEA_RECOVERY_CONTEXT;
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_WHEA_ERROR_NOTIFICATION)(
+    _In_ PWHEA_RECOVERY_CONTEXT RecoveryContext,
+    _In_ BOOLEAN PlatformDirected,
+    _In_ BOOLEAN Poisoned
+);
+
+typedef
+ULONG
+(NTAPI *PHAL_GET_PROCESSOR_INDEX_FROM_VP_INDEX)(
+    _In_ ULONG VpIndex
+);
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_SYNTHETIC_CLUSTER_IPI)(
+    _In_ PKAFFINITY_EX Affinity,
+    _In_ ULONG Vector
+);
+
+typedef
+BOOLEAN
+(NTAPI *PHAL_VP_START_ENABLED)(
+    VOID
+);
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_START_VIRTUAL_PROCESSOR)(
+    _In_ ULONG VpIndex,
+    _In_ PVOID Context
+);
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_GET_VP_INDEX_FROM_APIC_ID)(
+    _In_ ULONG ApicId,
+    _Out_ PULONG VpIndex
+);
+
+typedef struct _HAL_HV_SVM_SYSTEM_CAPABILITIES
+{
+    struct {
+        ULONG SvmSupported : 1;
+        ULONG GpaAlwaysValid : 1;
+    };
+    ULONG MaxPasidSpaceCount;
+    ULONG MaxPasidSpacePasidCount;
+    ULONG MaxPrqSize;
+    ULONG IommuCount;
+    ULONG MinIommuPasidCount;
+} HAL_HV_SVM_SYSTEM_CAPABILITIES, *PHAL_HV_SVM_SYSTEM_CAPABILITIES;
+
+typedef
+VOID
+(NTAPI *PHAL_SVM_GET_SYSTEM_CAPABILITIES)(
+    _Out_ PHAL_HV_SVM_SYSTEM_CAPABILITIES Capabilities
+);
+
+typedef
+SIZE_T
+(NTAPI *PHAL_IUM_EFI_RUNTIME_SERVICE)(
+    _In_ ULONG Service,
+    _Inout_updates_bytes_(Size) PVOID Data,
+    _In_ ULONGLONG Size,
+    _Inout_opt_ ULONGLONG Parameters[]
+);
+
+typedef struct _HAL_HV_SVM_DEVICE_CAPABILITIES
+{
+    struct {
+        ULONG SvmSupported : 1;
+        ULONG PciExecute : 1;
+        ULONG NoExecute : 1;
+        ULONG Reserved : 28;
+        ULONG OverflowPossible : 1;
+    };
+    ULONG PasidCount;
+    ULONG IommuIndex;
+} HAL_HV_SVM_DEVICE_CAPABILITIES, *PHAL_HV_SVM_DEVICE_CAPABILITIES;
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_SVM_GET_DEVICE_CAPABILITIES)(
+    _In_ ULONG DeviceId,
+    _Out_ PHAL_HV_SVM_DEVICE_CAPABILITIES Capabilities
+);
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_SVM_CREATE_PASID_SPACE)(
+    _In_ ULONG PasidSpaceId,
+    _In_ ULONG PasidCount
+);
+
+typedef
+NTSTATUS
+(NTAPI *PHAL_SVM_SET_PASID_ADDRESS_SPACE)(
+    _In_ ULONG PasidSpaceId,
+    _In_ ULONG Pasid,
+    _In_ ULONGLONG AddressSpace
+);
+
+typedef
+VOID
+(NTAPI *PHAL_SVM_FLUSH_PASID)(
+    _In_ ULONG PasidSpaceId,
+    _In_ ULONG Pasid,
+    _In_ ULONG Number,
+    _In_reads_opt_(Number) KTB_FLUSH_VA Virtual[]
```

Comment:
What's the difference between `_In_reads_` (opt in this case), and `_In_count_` that you use for the other instances?